File: python/src/machine_config/capabilities/merge.py
"""Merge / Replace helpers for facade set* APIs."""
from __future__ import annotations

import copy
from typing import Any

from machine_config.capabilities.generated import SetMode


def snapshot(value: Any) -> Any:
    return copy.deepcopy(value)


def apply_set_mode(
    current: dict[str, Any],
    incoming: dict[str, Any],
    mode: SetMode = SetMode.MERGE,
) -> dict[str, Any]:
    if mode == SetMode.REPLACE:
        return snapshot(incoming)
    out = snapshot(current)
    for key, value in incoming.items():
        out[key] = value
    return out

File: rust/src/capabilities/merge.rs
//! Merge / Replace for facade set* APIs.

use crate::capabilities::errors::CapabilityError;
use crate::capabilities::generated::SetMode;
use serde::de::DeserializeOwned;
use serde::Serialize;

pub fn apply_set_mode<T>(
    current: &T,
    incoming: &T,
    mode: SetMode,
) -> Result<T, CapabilityError>
where
    T: Serialize + DeserializeOwned + Clone,
{
    match mode {
        SetMode::Replace => Ok(incoming.clone()),
        SetMode::Merge => {
            let mut cur = serde_json::to_value(current)
                .map_err(|e| CapabilityError::validation_error(e.to_string()))?;
            let inc = serde_json::to_value(incoming)
                .map_err(|e| CapabilityError::validation_error(e.to_string()))?;
            if let (serde_json::Value::Object(c), serde_json::Value::Object(i)) = (&mut cur, inc)
            {
                for (k, v) in i {
                    c.insert(k, v);
                }
            }
            serde_json::from_value(cur)
                .map_err(|e| CapabilityError::validation_error(e.to_string()))
        }
    }
}
